File: Algorithm/Baekjoon/23년도/23_01_week_5/2240.cpp
#include <iostream>
#include <vector>

using namespace std;

int dp[32][1002];

int main()
{

    int T, W;
    cin >> T >> W;

    for (int i = 1; i <= T; i++)
    {
        int t_n;
        cin >> t_n;
        for (int j = 1; j <= W + 1 && j <= i + 1; j++)
        {
            if (t_n % 2 == j % 2)
            {
                if (dp[j][i - 1] + 1 > dp[j - 1][i - 1] + 1)
                {
                    dp[j][i] += dp[j][i - 1] + 1;
                }
                else
                {
                    dp[j][i] += dp[j - 1][i - 1] + 1;
                }
            }
            else
            {
                if (dp[j][i - 1] > dp[j - 1][i - 1])
                {
                    dp[j][i] += dp[j][i - 1];
                }
                else
                {
                    dp[j][i] += dp[j - 1][i - 1];
                }
            }
        }
    }

    int ans = 0;
    for (int i = 0; i <= W + 1; i++)
    {
        if (ans < dp[i][T])
        {
            ans = dp[i][T];
        }
    }

    cout << ans;

    return 0;
}

File: Algorithm/Baekjoon/23년도/23_01_week_5/2240.py
import sys

def input():
    return sys.stdin.readline().rstrip()

T, W = map(int, input().split())

dp = [[0]*(T+1) for _ in range(W+2)]

for i in range(1, T+1):
    t_n = int(input())
    for j in range(1, min(W+2, i+2)):
        if t_n % 2 == j % 2:
            dp[j][i] += max(dp[j][i-1] + 1, dp[j-1][i-1] + 1)
        else:
            dp[j][i] += max(dp[j][i-1], dp[j-1][i-1])
ans = 0
for lst in dp:
    ans = max(lst[-1], ans)
print(ans)
        
